TASK: Routing: Throw exception for optional route parts without default

bwaidelich opened the pull request:
In order to match/resolve unambiguously for every dynamic optional route
part a corresponding default value has to be specified:

```
-
  uriPattern: 'foo/({bar})'
  defaults:
    'bar': 'without-this-it-wont-work'
```

With this change an exception is thrown when a route with optional dynamic
parts without corresponding defaults is configured.

Resolves: #683

kitsunet commented:
Seems some test nicely exposes the new behavior and needs adaption...

bwaidelich commented:
> Seems some test nicely exposes the new behavior and needs adaption...

I don't understand?

bwaidelich commented on TYPO3.Flow/Tests/Unit/Mvc/Routing/RouteTest.php:
```diff
@@ -559,8 +559,9 @@ public function routeMatchesRequestPathWithAllPartsIfUriPatternContainsTwoOption
 
     /**
      * @test
+     * @expectedException \TYPO3\Flow\Mvc\Exception\InvalidRouteSetupException
```
This is new. With this adjusted test we now test the new behavior

bwaidelich commented on TYPO3.Flow/Tests/Unit/Mvc/Routing/RouteTest.php:
```diff
@@ -581,16 +582,6 @@ public function routeMatchesEmptyRequestPathIfUriPatternContainsOneOptionalDynam
     /**
      * @test
      */
-    public function routeDoesNotMatchRequestPathContainingNoneOfTheOptionalRoutePartsIfNoDefaultsAreSet()
```
These don't make sense any longer because this case can't ever happen

bwaidelich commented:
@hhoechtl @kitsunet *ping*

kitsunet commented:
Sorry, I restarted tests, but they are still red, I am unsure about your explanation. Shouldn't the tests all run?

bwaidelich commented:
> Shouldn't the tests all run?

Yes, they should. The failing travis CI is unrelated:
![image](https://cloud.githubusercontent.com/assets/307571/20009149/ae89c8d0-a2a3-11e6-8141-5bfe3ffad5fb.png)

kitsunet commented:
Huh, but that is just the line that says that there is a test failure. I see this:

```
1) TYPO3\Flow\Tests\Functional\Mvc\ActionControllerTest::defaultActionSpecifiedInrouteIsCalledAndResponseIsReturned

Failed asserting that two strings are equal.

--- Expected

+++ Actual

@@ @@

-'First action was called'

+'

+Uncaught Exception in Flow #1477140679: There is no default value specified for the optional route part "{@action}" of route "testb", but all dynamic optional route parts need a default.

+thrown in file /home/travis/build/neos/flow-development-distribution/Data/Temporary/Testing/Cache/Code/Flow_Object_Classes/TYPO3_Flow_Mvc_Routing_Route.php

+in line 643

+
```

bwaidelich commented:
@kitsunet doh, I must be blind ;)
I fixed the test now and rebased the PR (the failing tests are now really unrelated).
PS: This PR can be squashed on merge

albe commented:
Hm, the test failue still states

```
There was 1 failure:
1) TYPO3\Flow\Tests\Functional\Mvc\ViewsConfiguration\ViewsConfigurationTest::templatePathAndFilenameIsChanged
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'Changed on Package Level'
+'
+Uncaught Exception in Flow #1477140679: There is no default value specified for the optional route part "{@action}" of route "viewsconfigurationa", but all dynamic optional route parts need a default.
+thrown in file /home/travis/build/neos/flow-development-distribution/Data/Temporary/Testing/Cache/Code/Flow_Object_Classes/TYPO3_Flow_Mvc_Routing_Route.php
+in line 643
```

Still related I think ;)

bwaidelich commented:
uhm, this was getting embarrassing.. now for good! ;)

albe commented:
No worries, merges always carry possibility of (re)introducing wrong behaviour :) Looks good now
